### PR TITLE
Add TextEdits support to InlayHints

### PR DESCRIPTION
**Pull request by JoeRobich**

### CODE REVIEW

File: src/OmniSharp.Abstractions/Models/v1/InlayHints/InlayHint.cs [hunk 1]
```diff
@@ -6,30 +6,88 @@ namespace OmniSharp.Models.v1.InlayHints;
...
     }
 
-    public override int GetHashCode() => (Position, Label, Tooltip).GetHashCode();
+    public override int GetHashCode() => (Position, Label, Tooltip, TextEdits?.GetHashCode() ?? 0).GetHashCode();
 }
 
 public enum InlayHintKind
 {
+    /// <summary>
+    /// An inlay hint that for a type annotation.
```
Comment (333fred) [inline src/OmniSharp.Abstractions/Models/v1/InlayHints/InlayHint.cs:86]:
```suggestion
    /// An inlay hint that is for a type annotation.
```

File: src/OmniSharp.Roslyn.CSharp/Services/InlayHints/InlayHintService.cs [hunk 2]
```diff
@@ -152,6 +155,27 @@ public List<InlayHint> MapAndCacheHints(ImmutableArray<OmniSharpInlineHint> rosl
...
+        {
+            if (textChange.HasValue == false)
+            {
+                return Array.Empty<LinePositionSpanTextChange>();
+            }
+
+            var changeLinePositionSpan = sourceText.Lines.GetLinePositionSpan(textChange.Value.Span);
+            return new[]
+            {
+                new LinePositionSpanTextChange()
```
Comment (333fred) [inline src/OmniSharp.Roslyn.CSharp/Services/InlayHints/InlayHintService.cs:168]: Do we not have a helper for this conversion?

**Review by 333fred — APPROVED** [review]
LGTM, just a couple minor comments.

File: src/OmniSharp.Roslyn.CSharp/Services/InlayHints/InlayHintService.cs (hunk 2 above)
Comment (JoeRobich) [reply src/OmniSharp.Roslyn.CSharp/Services/InlayHints/InlayHintService.cs:168]: CompletionService uses something similar `CompletionListBuilder.GetChangeForTextAndSpan()`. I will move this bit to the Helpers class and update both to use it.

Comment (JoeRobich) [reply src/OmniSharp.Roslyn.CSharp/Services/InlayHints/InlayHintService.cs:168]: Should have looked harder, I just found the TextChangeHelper class.

**Comment by JoeRobich** [comment]
@bjorkstromm Seeing an odd Cake completion failure in the Windows net6.0 tests. The override completion is returning AdditionalTextEdits which appear to be formatting the DocComments from [DotNetCoreAliases.cs](https://github.com/cake-build/cake/blob/0e73b358b1f766f65feb2d5a0dc435ae77726d27/src/Cake.Common/Tools/DotNetCore/DotNetCoreAliases.cs). I saw the CompletionResolve handler was nulling out the AdditionalTextEdits property and was unsure if the CompletionRequest handler should be doing something similar. Surprised we only see this on Windows.

```
[xUnit.net 00:00:57.29]     OmniSharp.Cake.Tests.CompletionFacts.ShouldGetAdditionalTextEditsFromOverrideCompletion [FAIL][xUnit.net 00:00:57.29]       Assert.All() Failure: 5 out of 5 items in the collection did not pass.
[xUnit.net 00:00:57.29]       [4]: Item: { Label = ToString(), CompletionItemKind = Method }
[xUnit.net 00:00:57.29]            Xunit.Sdk.NullException: Assert.Null() Failure
[xUnit.net 00:00:57.29]            Expected: (null)
[xUnit.net 00:00:57.29]            Actual:   List<LinePositionSpanTextChange> [LinePositionSpanTextChange { StartLine=4155, StartColumn=7, EndLine=4203, EndColumn=7, NewText= }, LinePositionSpanTextChange { StartLine=4203, StartColumn=30, EndLine=4203, EndColumn=101, NewText= }, LinePositionSpanTextChange { StartLine=4207, StartColumn=76, EndLine=4207, EndColumn=86, NewText= }, LinePositionSpanTextChange { StartLine=4212, StartColumn=15, EndLine=4232, EndColumn=4, NewText=[deprecated] DotNetCoreRestore is obsolete and will be removed in a future release. Use <see cref="M:Cake.Common.Tools.DotNet.DotNetAliases.DotNetRestore(Cake.Core.ICakeContext,System.String)" /> instead.\r\n///            Restore all NuGet Packages in the specified path.\r\n///            </summary>\r\n/// <param name="context">The context.</param>\r\n/// <param name="root">Path to the project file to restore.</param>\r\n/// <example>\r\n///            <code>\r\n///            DotNetCoreRestore("./src/MyProject/MyProject.csproj");\r\n///            </code>\r\n/// }, LinePositionSpanTextChange { StartLine=4233, StartColumn=48, EndLine=4233, EndColumn=121, NewText= }, ...]
[xUnit.net 00:00:57.29]               at Xunit.Assert.Null(Object object) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\NullAsserts.cs:line 31
[xUnit.net 00:00:57.29]               at OmniSharp.Cake.Tests.CompletionFacts.<>c.<ShouldGetAdditionalTextEditsFromOverrideCompletion>b__9_3(CompletionItem c) in D:\a\omnisharp-roslyn\omnisharp-roslyn\tests\OmniSharp.Cake.Tests\CompletionFacts.cs:line 168
[xUnit.net 00:00:57.29]               at Xunit.Assert.All[T](IEnumerable`1 collection, Action`1 action) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\CollectionAsserts.cs:line 36
... Other completion items carry the same set of AdditionalTextEdits ...
```

**Comment by bjorkstromm** [comment]
> @bjorkstromm Seeing an odd Cake completion failure in the Windows net6.0 tests. The override completion is returning AdditionalTextEdits which appear to be formatting the DocComments from [DotNetCoreAliases.cs](https://github.com/cake-build/cake/blob/0e73b358b1f766f65feb2d5a0dc435ae77726d27/src/Cake.Common/Tools/DotNetCore/DotNetCoreAliases.cs). I saw the CompletionResolve handler was nulling out the AdditionalTextEdits property and was unsure if the CompletionRequest handler should be doing something similar. Surprised we only see this on Windows.
> 
> ```
> [xUnit.net 00:00:57.29]     OmniSharp.Cake.Tests.CompletionFacts.ShouldGetAdditionalTextEditsFromOverrideCompletion [FAIL][xUnit.net 00:00:57.29]       Assert.All() Failure: 5 out of 5 items in the collection did not pass.
> [xUnit.net 00:00:57.29]       [4]: Item: { Label = ToString(), CompletionItemKind = Method }
> [xUnit.net 00:00:57.29]            Xunit.Sdk.NullException: Assert.Null() Failure
> [xUnit.net 00:00:57.29]            Expected: (null)
> [xUnit.net 00:00:57.29]            Actual:   List<LinePositionSpanTextChange> [LinePositionSpanTextChange { StartLine=4155, StartColumn=7, EndLine=4203, EndColumn=7, NewText= }, LinePositionSpanTextChange { StartLine=4203, StartColumn=30, EndLine=4203, EndColumn=101, NewText= }, LinePositionSpanTextChange { StartLine=4207, StartColumn=76, EndLine=4207, EndColumn=86, NewText= }, LinePositionSpanTextChange { StartLine=4212, StartColumn=15, EndLine=4232, EndColumn=4, NewText=[deprecated] DotNetCoreRestore is obsolete and will be removed in a future release. Use <see cref="M:Cake.Common.Tools.DotNet.DotNetAliases.DotNetRestore(Cake.Core.ICakeContext,System.String)" /> instead.\r\n///            Restore all NuGet Packages in the specified path.\r\n///            </summary>\r\n/// <param name="context">The context.</param>\r\n/// <param name="root">Path to the project file to restore.</param>\r\n/// <example>\r\n///            <code>\r\n///            DotNetCoreRestore("./src/MyProject/MyProject.csproj");\r\n///            </code>\r\n/// }, LinePositionSpanTextChange { StartLine=4233, StartColumn=48, EndLine=4233, EndColumn=121, NewText= }, ...]
> [xUnit.net 00:00:57.29]               at Xunit.Assert.Null(Object object) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\NullAsserts.cs:line 31
> [xUnit.net 00:00:57.29]               at OmniSharp.Cake.Tests.CompletionFacts.<>c.<ShouldGetAdditionalTextEditsFromOverrideCompletion>b__9_3(CompletionItem c) in D:\a\omnisharp-roslyn\omnisharp-roslyn\tests\OmniSharp.Cake.Tests\CompletionFacts.cs:line 168
> [xUnit.net 00:00:57.29]               at Xunit.Assert.All[T](IEnumerable`1 collection, Action`1 action) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\CollectionAsserts.cs:line 36
> ... Other completion items carry the same set of AdditionalTextEdits ...
> ```

@JoeRobich, hmm unsure what the correct approach is. I'll have a look soon. But I think nulling might be the correct approach. Ok if I push to your branch if I find a solution?

**Comment by bjorkstromm** [comment]
@JoeRobich, I checked the code. 

So, previously (if my old code comment was correct) `AdditionalTextEdits` didn't include incremental changes, but instead the whole buffer, which made it problematic for Cake (since DSL is part of the document). The correct fix would be to translate lines in the `AdditionalTextEdits` so that generated DSL is not part of the response. I can do this later. Quick fix is to just make it `null` like in the resolve handler. I can update your branch soon.

Did this PR change behavior so that `AdditionalTextEdits` now is returned in the completion request, and not in the completion resolve request? If not, it's weird that we haven't seen this before.

**Comment by JoeRobich** [comment]
> Quick fix is to just make it null like in the resolve handler. I can update your branch soon.
I'll give that a shot.

> Did this PR change behavior so that AdditionalTextEdits now is returned in the completion request, and not in the completion resolve request? If not, it's weird that we haven't seen this before.
No intentional behavior changes for completion, so it is odd indeed.

**Comment by bjorkstromm** [comment]
A closer look at the code, and I see that we currently are translating line numbers for the `AdditionalTextEdits`, thus edits for DSL should be removed.
https://github.com/OmniSharp/omnisharp-roslyn/blob/04a561ca69584ab0bad1fd3fa75a14006e9afd7d/src/OmniSharp.Cake/Extensions/ResponseExtensions.cs#L214

I was however unable to reproduce locally on Windows. I'd say go ahead and make them `null` still. It "shouldn't" break anything.

**Review by bjorkstromm — APPROVED** [review]

**Pull request merged**